File: src/main/java/org/openmedstack/Service.java
package org.openmedstack;

import org.openmedstack.commands.CommandResponse;
import org.openmedstack.commands.DomainCommand;
import org.openmedstack.events.BaseEvent;

import java.util.Arrays;
import java.util.concurrent.CompletableFuture;

public interface Service extends AutoCloseable {
    CompletableFuture start();

    <T extends DomainCommand> CompletableFuture<CommandResponse> send(T command);

    <T extends BaseEvent> CompletableFuture publish(T msg);

    <T extends Object> T resolve(Class<T> type);
}

File: src/main/kotlin/org/openmedstack/Service.kt
package org.openmedstack

import org.openmedstack.commands.CommandResponse
import org.openmedstack.commands.DomainCommand
import org.openmedstack.events.BaseEvent
import java.util.concurrent.CompletableFuture

interface Service : AutoCloseable {
    fun start(): CompletableFuture<*>
    fun <T> send(command: T): CompletableFuture<CommandResponse> where T : DomainCommand
    fun <T> publish(msg: T): CompletableFuture<*> where T : BaseEvent
    fun <T> resolve(type: Class<T>): T where T : Any
}
